modules/solid_mechanics/src/kernels/StressDivergenceRZ.C

```diff
@@ -101,11 +227,103 @@ StressDivergenceRZ::calculateJacobian( unsigned int ivar, unsigned int jvar )
   }
 
   SymmTensor tmp( _Jacobian_mult[_qp] * phi );
-  const Real val( test.doubleContraction( tmp ) );
+  Real val( test.doubleContraction( tmp ) );
 
+  // volumetric locking correction
+  // K = Bbar^T_i * C * Bbar^T_j where Bbar = B + Bvol
+  // K = B^T_i * C * B_j + Bvol^T_i * C * Bvol_j + B^T_i * C * Bvol_j + Bvol^T_i * C * B_j
+  if (_volumetric_locking_correction)
+  {
+    RealGradient new_test(2, 0.0);
+    RealGradient new_phi(2, 0.0);
+
+    new_test(0) = _grad_test[_i][_qp](0) + _test[_i][_qp] / _q_point[_qp](0);
+    new_test(1) = _grad_test[_i][_qp](1);
+    new_phi(0) = _grad_phi[_j][_qp](0) + _phi[_j][_qp] / _q_point[_qp](0);
+    new_phi(1) = _grad_phi[_j][_qp](1);
+
+    // Bvol^T_i * C * Bvol_j
+    val += _Jacobian_mult[_qp].sum_3x3() * (_avg_grad_test[_i][ivar] - new_test(ivar)) * (_avg_grad_phi[_j][jvar] - new_phi(jvar)) / 9.0;
+
+    // B^T_i * C * Bvol_j
+    RealGradient sum_3x1 = _Jacobian_mult[_qp].sum_3x1();
+    if ((ivar == 0) && (jvar == 0))
+      val += (sum_3x1(0) * test.xx() +  sum_3x1(2) * test.zz()) * (_avg_grad_phi[_j][0] - new_phi(0)) / 3.0;
+    else if ((ivar == 0) && (jvar == 1))
+      val += (sum_3x1(0) * test.xx() +  sum_3x1(2) * test.zz()) * (_avg_grad_phi[_j][1] - new_phi(1)) / 3.0;
+    else if ((ivar == 1) && (jvar == 0))
+      val += sum_3x1(1) * test.yy() * (_avg_grad_phi[_j][0] - new_phi(0)) / 3.0;
+    else
+      val += sum_3x1(1) * test.yy() * (_avg_grad_phi[_j][1] - new_phi(1)) / 3.0;
+
+    // Bvol^T_i * C * B_j
+    // tmp = C * B_j from above
+    if (ivar == 0)
+      val += (tmp.xx() + tmp.yy() + tmp.zz()) * (_avg_grad_test[_i][0] - new_test(0)) / 3.0;
+    else
+      val += (tmp.xx() + tmp.yy() + tmp.zz()) * (_avg_grad_test[_i][1] - new_test(1)) / 3.0;
+  }
   return val;
 }
 
+void
+StressDivergenceRZ::computeOffDiagJacobian(unsigned int jvar)
+{
+  if (jvar == _var.number())
+    computeJacobian();
+  else
+  {
+    if (_volumetric_locking_correction)
+    {
+      Real vol = 0.0;
+      // calculate volume averaged value of shape function derivative
+      _avg_grad_test.resize(_test.size());
+      for (_i = 0; _i < _test.size(); _i++)
```

Comment:
would you mind switching to pre-increment in for loops everywhere? (`; ++_i)`). It's never slower and sometimes faster (plus it is semantically what you _really_ mean)